src/sleep/sleep.c

```diff
@@ -264,78 +264,117 @@ static int execute(
-                else
-                        return log_error_errno(r, "Error fetching battery capacity percentage: %m");
-
-                r = get_total_suspend_interval(last_capacity, &total_suspend_interval);
-                if (r < 0)
-                        log_debug_errno(r, "Failed to estimate suspend interval using previous discharge rate, ignoring: %m");
-                else
-                        suspend_interval = total_suspend_interval;
+        r = freeze_thaw_user_slice(&method_freeze);
+        if (r < 0) {
+                log_debug_errno(r, "Failed to freeze unit user.slice, ignoring: %m");
+                method = NULL;
+        }
 
-                log_debug("Set timerfd wake alarm for %s", FORMAT_TIMESPAN(suspend_interval, USEC_PER_SEC));
-                /* Wake alarm for system with or without battery to hibernate or estimate discharge rate whichever is applicable */
-                timespec_store(&ts.it_value, suspend_interval);
+        r = check_wakeup_type(&wakeup_type_bit);
+        if (r < 0)
+                log_debug_errno(r, "Failed to fetch wakeup-type bit, ignoring: %m");
 
-                if (timerfd_settime(tfd, 0, &ts, NULL) < 0)
-                        return log_error_errno(errno, "Error setting battery estimate timer: %m");
+        r = battery_trip_point_alarm_exists();
+        if (r < 0)
+                log_debug_errno(r, "Failed to check acpi_btp support enabled or not, ignoring: %m");
 
+        if (r > 0 && wakeup_type_bit > 0) {
+                log_debug("Attempting to suspend...");
                 r = execute(sleep_config, SLEEP_SUSPEND, NULL);
                 if (r < 0)
                         return r;
 
-                r = fd_wait_for_event(tfd, POLLIN, 0);
+                r = check_wakeup_type(&wakeup_type_bit);
                 if (r < 0)
-                        return log_error_errno(r, "Error polling timerfd: %m");
-                /* Store fd_wait status */
-                woken_by_timer = FLAGS_SET(r, POLLIN);
+                        return log_debug_errno(r, "Failed to fetch wakeup-type bit: %m");
 
-                r = fetch_batteries_capacity_by_name(&current_capacity);
-                if (r < 0) {
-                        /* In case of no battery or error while getting charge level, no need to measure
-                         * discharge rate. Instead system should wakeup if it is manual wakeup or
-                         * hibernate if this is a timer wakeup.   */
-                        log_debug_errno(r, "Battery capacity percentage unavailable, cannot estimate discharge rate: %m");
-                        if (!woken_by_timer)
-                                return 0;
-                        break;
-                }
+                if (wakeup_type_bit != 3)
+                        /* For APM Timer wakeup, system should hibernate else wakeup */
+                        return 0;
+        } else
+                while (battery_is_low() == 0) {
```

Comment:
This `while` loop should be a separate function.

Reply:
Could be but avoided doing that since we need to iterate over suspend with the timer here based on different helper results and then executing SUSPEND or HIBERNATE so code seemed more readable and understandable to me that way....

Reply:
@keszybz I have fixed that and pushed in acpi_btp PR, here. https://github.com/systemd/systemd/pull/23895
Once that is merged after review, i will rebase here. that should reflect here then.